docs/decisions/0005-lti-provider-1.3-spec-support.rst

```diff
@@ -0,0 +1,651 @@
+   # scores back later. We know that the consumer exists, since the record was
+   # used earlier to verify the oauth signature.
+   store_outcome_parameters(params, request.user, lti_consumer)
+
+And finally, it will use the usage_key to return back a HttpResponse
+with the XBlock using ``render_courseware`` view.
+
+LTI Provider Outcome Service
+---------------------------
+
+The LTI provider is able to pass grades back to the campus LMS platform
+using the LTI outcome service. For full details of the outcome service, see:
+http://www.imsglobal.org/LTI/v1p1/ltiIMGv1p1.html
+
+In brief, the LTI 1.1 spec defines an outcome service that can be offered
+by an LTI consumer. The consumer determines whether a score should be
+returned (in Canvas, this means that the LTI tool is used in an assignment,
+and the launch was performed by a student). If so, it sends two additional
+parameters along with the LTI launch:
+
+- lis_outcome_service_url: the endpoint for the outcome service on the consumer;
+- lis_result_sourcedid: a unique identifier for the row in the gradebook.
+
+The LTI Provider launch view detects the presence of these optional fields,
+and creates database records for the specific Outcome Service and for the
+graded LTI launch.
+
+.. code:: python
+
+   # Store any parameters required by the outcome service in order to report
+   # scores back later. We know that the consumer exists, since the record was
+   # used earlier to verify the oauth signature.
+   store_outcome_parameters(params, request.user, lti_consumer)
+
+.. code:: python
+   # Create a record of the outcome service if necessary
+   outcomes, __ = OutcomeService.objects.get_or_create(
+      lis_outcome_service_url=result_service,
+      lti_consumer=lti_consumer
+   )
+
+   GradedAssignment.objects.get_or_create(
+      lis_result_sourcedid=result_id,
+      course_key=course_key,
+      usage_key=usage_key,
+      user=user,
+      outcome_service=outcomes
+   )
+
+Later, when a score on edX changes (identified using the signal mechanism):
+
+.. code:: python
+   @receiver(grades_signals.PROBLEM_WEIGHTED_SCORE_CHANGED)
+   def score_changed_handler(sender, **kwargs):  # pylint: disable=unused-argument
+      """
+      Consume signals that indicate score changes. See the definition of
+      PROBLEM_WEIGHTED_SCORE_CHANGED for a description of the signal.
+      """
+
+While handling the score change, first it will get all the assignments related
+to the course_key and usage_key received from the signal, and increment each one
+version_number by 1, this version number is used to avoid race conditions
+while sending score updates:
+
+.. code:: python
+   # Get all assignments involving the current problem for which the campus LMS
+   # is expecting a grade. There may be many possible graded assignments, if
+   # a problem has been added several times to a course at different
+   # granularities (such as the unit or the vertical).
```

Comment:
I know the documentation for get_assignments_for_problem implies that a single problem can exist multiple times in a course, and at different granularities, but I'm pretty sure that's not true.  If you _copy_ a problem component you end up with a new problem component with its own block id (usage_key). So multiple _copies_ of a problem can certainly be made, and located either in the same assignment/subsection (but why would you do this unless using cohort or track-based differentiated content), or in multiple assignments, but in either case they will each be a distinct problem instance.  I don't believe (happy to be proven wrong!) that get _assignment_for_problem should return more than one assignment.  
Confusing I know but Open edX uses both 'unit' and 'vertical' to mean the same thing.  
Also, content only exists at the vertical/unit level.  You can't add content blocks of any type at any other level, at least not from Studio. Attempting to do so by manipulating the OLX does kinda work in that Studio will load the file and try to represent it, but on publish the LMS renders the component under an existing subsection.

Reply:
It think get_assignments_for_problem was made because PROBLEM_WEIGHTED_SCORE_CHANGED describes in it's source code that it's only sent on problem score changes: 

```
# Signal that indicates that a user's weighted score for a problem has been updated.
# This signal is generated when a scoring event occurs in the Submissions module
# or a PROBLEM_RAW_SCORE_CHANGED event is handled in the core platform.
```

That means that in order to update outcomes for LTI 1.1 launches made for vertical components, the receiver will need to catch the score changes on vertical components from it's children, currently get_assignments_for_problem does that by tracing the parent hierarchy of a given problem:

```
locations = []
current_descriptor = problem_descriptor
while current_descriptor:
    locations.append(current_descriptor.location)
    current_descriptor = current_descriptor.get_parent()
assignments = GradedAssignment.objects.filter(
    user=user_id, course_key=course_key, usage_key__in=locations
)
return assignments
```

Now, right now I'm not sure on what's the behavior of PROBLEM_WEIGHTED_SCORE_CHANGED signal, this signal only translates the information sent from the score_set signal defined in `submission.models`, score_set signal is described has always informing any score change from any item:

```
# Signal to inform listeners that a score has been changed
score_set = Signal(providing_args=[
    'points_possible', 'points_earned', 'anonymous_user_id',
    'course_id', 'item_id', 'created_at'
])
```

If that's the case, then you are right that there should only be one LTI assignment to be catch on score_set or `PROBLEM_WEIGHTED_SCORE_CHANGED` signal for a given user_id, course_key and usage_key. 

I will look at this on further detail since there are more details I'm missing from how grading is handled by the platform.